libcaf_core/caf/flow/observable.test.cpp

```diff
@@ -68,6 +68,106 @@ TEST("element_at(n) only takes the element with index n") {
   }
 }
 
+TEST("reduce(z, f) applies a function to each item, sequentially, and emits "
+     "the final value") {
+  SECTION("reduce(z, f) emits final value for a range of size m") {
+    SECTION("blueprint") {
+      check_eq(collect(range(1, 1).reduce(0, std::plus<int>())), vector{1});
+      check_eq(collect(range(1, 2).reduce(0, std::plus<int>())), vector{3});
+      check_eq(collect(range(1, 3).reduce(0, std::plus<int>())), vector{6});
+    }
+    SECTION("observable") {
+      check_eq(collect(mat(range(1, 1)).reduce(0, std::plus<int>())),
+               vector{1});
+      check_eq(collect(mat(range(1, 2)).reduce(0, std::plus<int>())),
+               vector{3});
+      check_eq(collect(mat(range(1, 3)).reduce(0, std::plus<int>())),
+               vector{6});
+    }
+  }
+  SECTION("reduce(z, f) emits 0 for a range of size 0") {
+    SECTION("blueprint") {
+      check_eq(collect(range(1, 0).reduce(0, std::plus<int>())), vector{0});
+    }
+    SECTION("observable") {
+      check_eq(collect(mat(range(1, 0)).reduce(0, std::plus<int>())),
+               vector{0});
+    }
+  }
+  SECTION("reduce(z, f) propagates errors") {
+    SECTION("blueprint") {
+      check_eq(collect(
+                 range(1, 1).concat(obs_error()).reduce(0, std::plus<int>())),
+               sec::runtime_error);
+      check_eq(collect(
+                 range(1, 3).concat(obs_error()).reduce(0, std::plus<int>())),
+               sec::runtime_error);
+      check_eq(collect(obs_error().reduce(0, std::plus<int>())),
+               sec::runtime_error);
+    }
+    SECTION("observable") {
+      check_eq(
+        collect(
+          mat(range(1, 1).concat(obs_error())).reduce(0, std::plus<int>())),
+        sec::runtime_error);
+      check_eq(
+        collect(
+          mat(range(1, 3).concat(obs_error())).reduce(0, std::plus<int>())),
+        sec::runtime_error);
+      check_eq(collect(mat(obs_error()).reduce(0, std::plus<int>())),
+               sec::runtime_error);
+    }
+  }
+}
+
+TEST("scan(z, f) applies a function to each item sequentially, and emits each "
+     "successive value") {
```

Comment:
```suggestion
TEST("scan(init, fn) creates successive reduced values using fn") {
```

(plus using the `adder` as mentioned before to reduce verbosity)